packages/peregrine/src/hooks/useRestResponse.js

```diff
@@ -0,0 +1,121 @@
+import { useCallback, useMemo, useReducer } from 'react';
+
+const initialState = {
+    data: null,
+    error: null,
+    loading: false
+};
+
+const reducer = (state, { payload, type }) => {
+    switch (type) {
+        case 'set data': {
+            return { ...state, data: payload };
+        }
+        case 'set error': {
```

Comment:
@jimbo should I just drop `set data` and `set error` from the API?  `useRestApi` uses `receive response` and `receive error`.

Reply:
`set data` and `set error` set those properties individually (ignoring `loading). Since we already had a `receive response` for successful responses (sets `data` to the payload, `error` to `null`, and `loading` to `false`, I wanted something similar for results that weren't `2xx`s.

So `receive error` sets `data` to null, `error` to the error message (payload), and `loading` to `false`.

Without this function we'd have to call each of `set data`, `set error`, and `set loading` individually.

I think having a `receive error` makes good sense, I'm just wondering if we need the individual setters mentioned above. I don't see them used anywhere, though that doesn't mean they aren't potentially useful. Ultimately I'm fine keeping them in.

Reply:
> Infact these seem really similar to me, maybe we can just use the one and call it useApiResult?

They are totally similar right now but I don't know that we want to combine them. The `useQueryResult` hook is pretty dependent on how apollo returns data, which I wouldn't want the REST API hook to have to care about.

Reply:
> The useQueryResult hook is pretty dependent on how apollo returns data

Is it? They look super similar to me with minor differences in how we handle errors. Even `useRestApi` and `useQuery` are pretty similar. You could do the payload manipulation within the `runQuery`/`sendRequest` functions and combine into a single `useApiResponse` hook.

Reply:
@sirugh We don't need to be overly aggressive about sharing abstractions. The priority here is ergonomics, so I'd like to leave some room for them to grow apart if necessary.

If they don't grow apart, and we find that they're doing the same thing, that will be a straightforward optimization in the future—especially since `useQueryResult` and `useRestResponse` are basically private.

Reply:
> should I just drop set data and set error from the API?
> [...] I don't see them used anywhere, though that doesn't mean they aren't potentially useful.

We could. I don't foresee using them, and it would save a few bytes. Perhaps we just don't define the methods, but we leave the reducer cases in?

Reply:
Could do that too, yeah. I'm fine with whatever honestly. Like you said, we can optimize later.

Reply:
@jimbo Yea, I talked to @supernova-at yesterday and we agreed that they may differ in the future so this is safer for the time being. If they merge later it won't affect the public API so it shouldn't be a big deal.